core/src/consensus.rs

```diff
@@ -642,6 +642,34 @@ impl Tower {
                     SwitchForkDecision::FailedSwitchDuplicateRollback(latest_duplicate_ancestor)
                 };
 
+                // `heaviest_subtree_fork_choice` entries are not cleaned by duplicate block purging/rollback logic,
+                // so this is safe to check here. We return here if the last voted slot was rolled back/purged due to
+                // being a duplicate because `ancestors`/`descendants`/`progress` structurs may be missing this slot due
+                // to duplicate purging. This would cause many of the `unwrap()` checks below to fail.
+                //
+                // TODO: Handle if the last vote is on a dupe, and then we restart. The dupe won't be in
```

Comment:
This todo feels pretty important, do you plan to follow up in another PR soon to address this?

Reply:
Yeah I think this and the lockout issue illustrated by this problem here: https://github.com/solana-labs/solana/pull/18160/files#diff-6d8458bb2e53158ac472a9ad4709e6a0a52b75d930c019013298e8acda133828R5166-R5167 require that the tower keep track of all the hashes of all the votes in the tower, and not just the slot number

This means making an ABI change to tower and a migration, it's kind of a PITA but needs to be done 🤕